### ISSUE TITLE
Suggestions

### ISSUE THREAD

**Issue by WojciechMigda** [dgarana/vulcano#63]
Hi,
what a nice project!
I only had a quick glance over what it can do, so maybe these are already implemented, but:
1. how about aliases to commands?
2. is it possible to run vulcano repl on a passed asyncio loop, just as it is possible with bare python_prompt_toolkit?
Thanks!

**Comment by dgarana** [comment]
Hi @WojciechMigda ,

Thanks for being interested and for the star ;)

Regarding the questions you made:

1. I've been thinking about adding aliases to commands, this is a nice feature for being able to execute the same command but writing less (which is really cool). For doing this change we'd need to check if this alias was previously registered (not really hard) and also register the function under this alias on Magma(the command manager). For now, Magma stores all the information related with the functions inside a dictionary (command_name_string (as key): Command_object(as value)), the command object stores everything related with the command itself (the function to execute, all arguments, help...etc) so, in order to implement this, we may need to change this dictionary to another object to manage those registered commands(and aliases), that shouldn't be super hard, but definitely we may lose some performance as we will need to add more logic into it.

2. That's a nice question, for now, there are no plans on adding support for being able to use asyncio for running the prompt. But definitely will take a look at the example below:
https://github.com/prompt-toolkit/python-prompt-toolkit/blob/master/examples/prompts/asyncio-prompt.py  Good catch

**Comment by WojciechMigda** [comment]
Thank you for the reply.
Another idea - is it possible to create commands that are dynamically enabled/disabled, e.g. using app config? For instance someone writes an app which allows to authenticate to the remote service and once authenticated some commands become available.

**Comment by dgarana** [comment]
This should be related with the ProfileManager, which is still per define. But this a good idea. For now, you could implement it using the app.context.
So let's say you have 1 command which is called login, for example:
```python
@app.command
def login(username, password):
    # handle the login request here
    req = requests.post('yoururl/login', data={'user': username, 'pass': password})
    ... etc ...
    # if it is ok: then set up the context:
    app.context['permissions'] = ['execute_method_1']

@app.command
def execute_method_1(arg_1, arg_2):
    if not 'permissions' in app.context:
        raise Exception('You must login before')
    if 'execute_method_1' not in app.context['permissions']:
        raise Exception('You don't have enough permissions for executing this command')
    # DO THE REST HERE
```

Of course this is a very hacky way of achieving this, what would be nice is being able to register, or not register functions dynamically, but some pieces will fail, as for example the lexer, as it is created at very beginning and not recreated dynamically.

Those are very nice ideas, could you create one feature request talking about this?

I'll add them to the next version milestone.

**Comment by WojciechMigda** [comment]
By 'feature request' you mean another 'issue'?

In the meantime, I have spotted few things with vulcano:
1. when the `user_input` is split into `command` and `arguments` there is no check for empty `user_input` (https://github.com/dgarana/vulcano/blob/03e75a75426c485e16a76d138f7f61ce8d295704/vulcano/app/classes.py#L138), so when the user just hits <ENTER> in the command line the app throws an exception.
2. I took the example provided on the main page and run it. I hit `h`, the autocomplete suggests three variants ("hi", "help", and "whoami"), and then I hit <SPACE>. It throws an exception.
3. I had problems with `inline_parser` getting stuck with the simplest string input. I checked `pyparsing` version I had installed and it was 2.2.0. I upgraded it to the latest (2.3.1) and the problem disappeared. Can you check if there should be a version check in prerequisites?

**Comment by WojciechMigda** [comment]
I've been in need of package like `vulcano` for a while now and the ideas I put here are actually features I thought about implementing if I myself would engage in writing it.
So, another one is pretty complex, but I leave it to your judgement if you would be willing to include it within `vulcano`.
I would find it useful to be able to have a sort of context/endpoint tree-like structure navigation, maybe with `cd`-like builtin command. Then commands' behaviour could depend on where in that context tree we are. For instance, if we do `cd drawer/closet` we can execute `pick socks`, that would not be possible anywhere else. Maybe even some with some argument inference, e.g. if we have an endpoint `/guest/{name}/` we could do `cd guest/Jerry` or `cd guest/Helen` and subsequent `hello` command could enable printing `Hello Jerry`, or `Hello Helen` because `name` would be inferred.

**Comment by dgarana** [comment]
Regarding the reported errors:
1. I've created an issue with a branch reproducing this error "bug/65", if you want to fix it yourself, you're welcome, otherwise, I'll fix it tomorrow. #65 

2. This should be fixed on issue #66

3. Totally right, this should be a requirement for the project. #67

------

About the ideas you have, 

This could be achived with "MasterCommands", which at the end, is just a decorator, so for example:
You could be able to register two kind of classes, those that are instantiated, and classes that will be instantiated within a context.
Imagine that you just registered this class as a MasterCommand:
```python
@app.master_command
class MyCloset_1(object):
     def pick_socks(self):
          return self.socks
```
By registering classes, you can access those methods by executing this on command line:
```
>>> mycloset_1 pick_socks
```

Or for example, by entering into this master command:
```
>>> mycloset_1
(mycloset_1)>>> pick_socks
```

this would be useful for preventing the user having to type everytime the full name of the master command, and of course we could think something about having a reserved context for those master commands.

**Comment by dgarana** [comment]
I've merged your change related with bug #65 into release/0.3.0.

**Comment by dgarana** [comment]
Also issue #67 should be solved

**Comment by dgarana** [comment]
Your changes were released in version 0.2.2. Thanks for your help 👍

**Comment by dgarana** [comment]
Going to close this issue. In case you need anything else related with features requested here, please feel free to open individual tasks.

Thanks.
